### PR TITLE
Distribute mesh divisions using Positions

### PR DESCRIPTION
**Pull request by GiudGiud**
refs #26223

### CODE REVIEW

**Comment by GiudGiud** [comment]
Umm one concern here is overlap between bins if the positions are not far away enough from each other. I think I ll add a check in debug

File: framework/src/meshdivisions/CartesianGridDivision.C [hunk 1]
```diff
@@ -21,8 +23,12 @@ CartesianGridDivision::validParams()
   params.addClassDescription("Divide the mesh along a Cartesian grid. Numbering increases from "
                              "bottom to top and from left to right and from back to front. The "
                              "inner ordering is X, then Y, then Z");
-  params.addRequiredParam<Point>("bottom_left", "Bottom-back-left corner of the grid");
-  params.addRequiredParam<Point>("top_right", "Top-front-right corner of the grid");
+  params.addParam<Point>("bottom_left", "Bottom-back-left corner of the grid");
+  params.addParam<Point>("top_right", "Top-front-right corner of the grid");
+  params.addParam<Point>("center", "Center of the Cartesian grid");
+  params.addParam<PositionsName>("center_positions",
+                                 "Positions of the centers of distributed Cartesian grids");
```
Comment (lindsayad) [inline framework/src/meshdivisions/CartesianGridDivision.C:30]:
```suggestion
                                 "Positions of the centers of divided Cartesian grids");
```

File: framework/src/meshdivisions/CartesianGridDivision.C [hunk 2]
```diff
@@ -77,16 +104,33 @@ CartesianGridDivision::divisionIndex(const Elem & elem) const
 unsigned int
 CartesianGridDivision::divisionIndex(const Point & pt) const
 {
+  unsigned int offset = 0;
+  // Determine the local grid bounds
+  Point bottom_left = _bottom_left, top_right = _top_right, p = pt;
```
Comment (lindsayad) [inline framework/src/meshdivisions/CartesianGridDivision.C:109]: I would put this in a following `else` condition. I think it helps communicate that this line isn't used `if (_center_positions)`

File: framework/src/meshdivisions/CartesianGridDivision.C [hunk 3]
```diff
@@ -77,16 +104,33 @@ CartesianGridDivision::divisionIndex(const Elem & elem) const
...
+  if (_center_positions)
+  {
+    // If distributing using positions, find the closest position and
```
Comment (lindsayad) [inline framework/src/meshdivisions/CartesianGridDivision.C:112]:
```suggestion
    // If dividing using positions, find the closest position and
```

File: framework/src/meshdivisions/CylindricalGridDivision.C [hunk 4]
```diff
@@ -117,8 +131,24 @@ CylindricalGridDivision::divisionIndex(const Point & pt) const
...
+  Point in_plane;
+  unsigned int offset = 0;
+  if (_center)
+  {
+    pc(2) = (pt - *_center) * _direction;
+    in_plane = (pt - *_center) - pc(2) * _direction;
+  }
+  else
+  {
+    // If distributing using positions, find the closest position
```
Comment (lindsayad) [inline framework/src/meshdivisions/CylindricalGridDivision.C:143]:
```suggestion
    // If dividing using positions, find the closest position
```

File: framework/src/meshdivisions/CylindricalGridDivision.C [hunk 5]
```diff
@@ -117,8 +131,24 @@ CylindricalGridDivision::divisionIndex(const Point & pt) const
...
+    bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
Comment (lindsayad) [inline framework/src/meshdivisions/CylindricalGridDivision.C:144]:
```suggestion
    const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

File: framework/src/meshdivisions/CartesianGridDivision.C [hunk 6]
```diff
@@ -77,16 +104,33 @@ CartesianGridDivision::divisionIndex(const Elem & elem) const
...
+    // look at the relative position of the point compared to that position
+    bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
Comment (lindsayad) [inline framework/src/meshdivisions/CartesianGridDivision.C:114]:
```suggestion
    const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

File: framework/src/meshdivisions/SphericalGridDivision.C [hunk 7]
```diff
@@ -77,7 +88,17 @@ SphericalGridDivision::divisionIndex(const Point & pt) const
...
   // Compute coordinates of the point in the spherical coordinates
   Point pc;
-  pc(0) = (pt - _center).norm();
+  unsigned int offset = 0;
+  if (_center)
+    pc(0) = (pt - *_center).norm();
+  else
+  {
+    // If distributing using positions, find the closest position
+    bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
Comment (lindsayad) [inline framework/src/meshdivisions/SphericalGridDivision.C:97]:
```suggestion
    const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

File: test/tests/meshdivisions/tests [hunk 8]
```diff
@@ -21,6 +21,22 @@
...
+      csvdiff = 'cartesian_division_out_div_out_0001.csv'
+      detail = 'a regular Cartesian grid specified using its center and width,'
+    []
+    [cartesian_center_pos]
+      type = CSVDiff
+      input = 'cartesian_division.i'
+      cli_args = "MeshDivisions/active='cartesian_div_center_pos' "
+                 "AuxKernels/mesh_div/mesh_division=cartesian_div_center_pos"
+      csvdiff = 'cartesian_division_out_div_out_0001.csv'
+      detail = 'distributed Cartesian grids using a fixed width and a group of center positions,'
```
Comment (lindsayad) [inline test/tests/meshdivisions/tests:38]:
```suggestion
      detail = 'Cartesian grids with a fixed width and a group of center positions,'
```

File: test/tests/meshdivisions/tests [hunk 9]
```diff
@@ -39,12 +55,26 @@
       csvdiff = 'cylindrical_division_out_div_out_0001.csv'
       detail = 'a regular cylindrical R-theta-Z grid,'
     []
+    [cylinder_center_positions]
+      type = CSVDiff
+      input = 'cylindrical_division.i'
+      cli_args = 'MeshDivisions/active=cylindrical_div_positions AuxKernels/mesh_div/mesh_division=cylindrical_div_positions'
+      csvdiff = 'cylindrical_division_out_div_out_0001.csv'
+      detail = 'distributed cylindrical R-theta-Z grids using a group of positions for the cylinder axis coordinates,'
```
Comment (lindsayad) [inline test/tests/meshdivisions/tests:63]:
```suggestion
      detail = 'cylindrical R-theta-Z grids with a group of positions for the cylinder axis coordinates,'
```

File: framework/src/meshdivisions/CartesianGridDivision.C (hunk 2 above)
Comment (GiudGiud) [reply framework/src/meshdivisions/CartesianGridDivision.C]: done

**Comment by GiudGiud** [comment]
Added block restriction to SUbdomainsDivision.
The true potential of all this will show in the transfers PR, where for the first time we will be able to match subdomains to subdomains in a Transfer. Or match subdomains to subapps. Both are pretty cool and only a special case of the divisions

File: framework/doc/content/source/meshdivisions/SubdomainsDivision.md [hunk 10]
```diff
@@ -2,9 +2,15 @@
 
 !syntax description /MeshDivisions/SubdomainsDivision
 
+If not setting the [!param](/MeshDivisions/SubdomainsDivision/block) parameter:
 The divisions are numbered from 0 to the number of blocks in the mesh minus one.
 The ordering of the divisions is the same of the ordering of the IDs of the subdomains.
 
+If specifying a block restriction using the [!param](/MeshDivisions/SubdomainsDivision/block) parameter:
+The divisions are numbered from 0 to the number of blocks specified. An invalid id is returned
```
Comment (lindsayad) [inline framework/doc/content/source/meshdivisions/SubdomainsDivision.md:10]: I would expect the math here to be similar. So 0 to the number of blocks specified minus one is what I would expect

File: framework/src/meshdivisions/SubdomainsDivision.C [hunk 11]
```diff
@@ -19,35 +19,55 @@ InputParameters
...
 {
-  // The subdomains may not be contiguously numbered
-  std::set<libMesh::subdomain_id_type> subdomain_ids;
-  _mesh.getMesh().subdomain_ids(subdomain_ids, /*global=*/true);
-  setNumDivisions(subdomain_ids.size());
+  if (!blockRestricted())
+  {
+    // The subdomains may not be contiguously numbered
+    std::set<libMesh::subdomain_id_type> subdomain_ids;
+    _mesh.getMesh().subdomain_ids(subdomain_ids, /*global=*/true);
```
Comment (lindsayad) [inline framework/src/meshdivisions/SubdomainsDivision.C:42]: You could just call `blockIDs()` here

File: framework/src/meshdivisions/SubdomainsDivision.C [hunk 12]
```diff
@@ -19,35 +19,55 @@ InputParameters
...
-    _subdomain_ids_to_division_index[sub_id] = i++;
+    setNumDivisions(subdomain_ids.size());
+    unsigned int i = 0;
+    for (const auto sub_id : subdomain_ids)
+      _subdomain_ids_to_division_index[sub_id] = i++;
+  }
+  else
+  {
+    // Follow the user input order, use a vector instead of a set
+    const auto subdomain_ids = _mesh.getSubdomainIDs(getParam<std::vector<SubdomainName>>("block"));
```
Comment (lindsayad) [inline framework/src/meshdivisions/SubdomainsDivision.C:52]: Could call `getSubdomainIDs` with `blocks()`

Could also create an API on `BlockRestrictable` that returns its `_vec_ids`

File: framework/src/meshdivisions/SubdomainsDivision.C [hunk 13]
```diff
@@ -19,35 +19,55 @@ InputParameters
...
+    for (const auto sub_id : subdomain_ids)
+      _subdomain_ids_to_division_index[sub_id] = i++;
+  }
 }
 
 unsigned int
 SubdomainsDivision::divisionIndex(const Elem & elem) const
 {
-  return libmesh_map_find(_subdomain_ids_to_division_index, elem.subdomain_id());
+  auto id = _subdomain_ids_to_division_index.find(elem.subdomain_id());
```
Comment (lindsayad) [inline framework/src/meshdivisions/SubdomainsDivision.C:63]: Would mark `const` and rename as `id_it`

**Comment by GiudGiud** [comment]
Did all 4. Thanks for the suggestions! It definitely looks better to use the API that's cloest to the need

File: framework/src/meshdivisions/SubdomainsDivision.C [hunk 14]
```diff
@@ -38,9 +38,7 @@ SubdomainsDivision::initialize()
   if (!blockRestricted())
   {
     // The subdomains may not be contiguously numbered
-    std::set<libMesh::subdomain_id_type> subdomain_ids;
-    _mesh.getMesh().subdomain_ids(subdomain_ids, /*global=*/true);
-
+    const auto subdomain_ids = blockIDs();
```
Comment (lindsayad) [inline framework/src/meshdivisions/SubdomainsDivision.C:41]: Maybe you would get copy elision here, but you can catch by reference with this one

Comment (GiudGiud) [reply framework/src/meshdivisions/SubdomainsDivision.C]: fixed

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
failure in NS is unrelated

thanks for the review!

**Pull request merged**